grpcx/server.go

```diff
@@ -56,22 +58,78 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
 
 // WithErrorLogs returns a gRPC interceptor for unary calls that instrument requests
 // with logs for the errors.
-func WithErrorLogs(l logx.Logger) grpc.UnaryServerInterceptor {
+func WithErrorLogs(l logx.Logger, options ...func(*WithErrorLogsOptions)) grpc.UnaryServerInterceptor {
+	var logOptions = &WithErrorLogsOptions{
+		debugLevelCodes: []codes.Code{},
+		infoLevelCodes:  []codes.Code{codes.Canceled, codes.Unknown, codes.InvalidArgument, codes.DeadlineExceeded, codes.NotFound, codes.AlreadyExists, codes.PermissionDenied, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange, codes.Unimplemented, codes.Internal, codes.Unavailable, codes.DataLoss, codes.Unauthenticated},
+	}
+	for _, option := range options {
+		option(logOptions)
+	}
+
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
 		resp, err := handler(ctx, req)
 
 		if err != nil {
+			errCode := status.Code(err)
 			reqMsg, _ := json.Marshal(req)
 			respMsg, _ := json.Marshal(resp)
 			fields := []logx.Field{
 				{Key: "ctx_full_method", Value: info.FullMethod},
 				{Key: "ctx_request_content", Value: string(reqMsg)},
 				{Key: "ctx_response_content", Value: string(respMsg)},
-				{Key: "ctx_response_error", Value: err.Error()},
+				{Key: "ctx_response_error_code", Value: errCode},
+				{Key: "ctx_response_error_message", Value: err.Error()},
+			}
+			if inArray(errCode, logOptions.debugLevelCodes) {
+				l.Debug("gRPC Error", fields...)
+			}
+			if inArray(errCode, logOptions.infoLevelCodes) {
+				l.Info("gRPC Error", fields...)
 			}
-			l.Info("gRPC Error", fields...)
 		}
 
 		return resp, err
 	}
 }
+
+type WithErrorLogsOptions struct {
+	debugLevelCodes []codes.Code
+	infoLevelCodes  []codes.Code
+}
+
+func inArray(needle codes.Code, haystack []codes.Code) bool {
```

Comment:
this naming is leaking php like hell

Reply:
😅  you're right! changed to `inCodeList`